src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,411 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::build_errors::Error as BuilderError;
+use crate::credentials::internal::sts_exchange::ClientAuthentication;
+use crate::headers_util::build_cacheable_headers;
+use crate::token::{CachedTokenProvider, Token, TokenProvider};
+use crate::token_cache::TokenCache;
+use crate::{BuildResult, Result};
+use gax::error::CredentialsError;
+use http::{Extensions, HeaderMap};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::time::{Duration, Instant};
+
+use super::dynamic::CredentialsProvider;
+use super::external_account_sources::url_sourced_account::UrlSourcedSubjectTokenProvider;
+use super::internal::sts_exchange::{ExchangeTokenRequest, STSHandler};
+use super::{CacheableResource, Credentials};
+
+const CLOUD_PLATFORM_SCOPE: &str = "https://www.googleapis.com/auth/cloud-platform";
+
+#[async_trait::async_trait]
+pub(crate) trait SubjectTokenProvider: std::fmt::Debug + Send + Sync {
+    /// Generate subject token that will be used on STS exchange.
+    async fn subject_token(&self) -> Result<String>;
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub(crate) struct CredentialSourceFormat {
+    #[serde(rename = "type")]
+    pub format_type: String,
+    pub subject_token_field_name: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub(crate) struct CredentialSourceHeaders {
+    #[serde(flatten)]
+    pub headers: HashMap<String, String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub(crate) struct ExecutableConfig {
+    pub command: String,
+    pub timeout_millis: Option<u32>,
+    pub output_file: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(untagged)]
+enum CredentialSource {
+    UrlSourced {
+        url: String,
+        headers: Option<CredentialSourceHeaders>,
+        format: Option<CredentialSourceFormat>,
+    },
+    File {
+        file: String,
+        format: Option<CredentialSourceFormat>,
+    },
+    Executable {
+        executable: ExecutableConfig,
+    },
+    Aws {
+        environment_id: String,
+        region_url: Option<String>,
+        regional_cred_verification_url: Option<String>,
+        cred_verification_url: Option<String>,
+        imdsv2_session_token_url: Option<String>,
+    },
+}
+
+#[async_trait::async_trait]
+impl SubjectTokenProvider for CredentialSource {
+    async fn subject_token(&self) -> Result<String> {
+        match self.clone() {
+            CredentialSource::UrlSourced {
+                url,
+                headers,
+                format,
+            } => {
+                let source = UrlSourcedSubjectTokenProvider {
```

Comment:
Every time subject token is needed, are we going to figure out which is the cred type and build a provider?

Reply:
yeah, with this implementation yes. That's a way that I found to get rid of the `dyn SubjectTokenProvider` stuff